src/storage-controller/src/lib.rs

```diff
@@ -390,23 +353,100 @@ where
             ))
         }
 
-        // Perform all stash writes in a single transaction, to minimize transaction overhead and
-        // the time spent waiting for stash.
+        // Acquire read holds at the CollectionsController to ensure that these
+        // collections are not dropped while we're still working with them. We
+        // will release these holds only when a collection is dropped.
+        //
+        // TODO: We should also acquire read holds for sub-source and store them
+        // in our state. We're not doing that right now because we don't get
+        // `DroppedIds` messages for them, so we wouldn't know when to clean up
+        // read holds. Plus, the dependencies go from main source to sub source,
+        // and the CollectionsController will maintain holds from the main
+        // source on the sub-source, meaning we transitively have a hold on them
+        // from these holds here.
+        let desired_ingestion_read_holds = collections
+            .iter()
+            .filter(|(_id, desc)| desc.cluster_id().is_some())
```

Comment:
I believe webhooks also have cluster_id's, if we intend to keep this separate from the webhook block below

Reply:
They don't, according to https://github.com/MaterializeInc/materialize/blob/4dda3335e26fb0c8f3de4afe7c597ecb45ee3c75/src/storage-client/src/controller.rs#L587